### PR TITLE
Idea: Unlock remote machine via SSH

### PR DESCRIPTION
**Pull request by Frederick888**
This is a proof of concept which contains a lot of ugly hacks and it's not intended to be merged.

I would like to use yubikey-full-disk-encryption to unlock a headless machine without the hassle of unplugging/plugging YubiKey. So inspired by the `encryptssh` hook from [mkinitcpio-utils](https://www.archlinux.org/packages/community/any/mkinitcpio-utils/) I'm using currently, here's a prototype that utilises SSH port forwarding to achieve this job.

##### Server-Side Requirements
1. `nc` from [gnu-netcat](https://www.archlinux.org/packages/extra/x86_64/gnu-netcat/)
0. `ss` from [iproute2](https://www.archlinux.org/packages/core/x86_64/iproute2/)
0. a patched [mkinitcpio-dropbear](https://github.com/Frederick888/makepkg-mkinitcpio-dropbear-patched) (contains [an upstream dss fix](https://github.com/grazzolini/mkinitcpio-dropbear/commit/3905a71c1d2a447cf4c0583da1115ab324dc7977) and a minor change to allow port forwarding)

dropbear (and netconf) needs to be pre-configured according to https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Remote_unlocking_(hooks:_netconf,_dropbear,_tinyssh,_ppp).

mkinitcpio hooks should look like:
```
HOOKS=(base udev autodetect keyboard keymap modconf block netconf dropbear ykfde filesystems fsck)
```

#### Client-Side Requirements
1. `ncat` from [nmap](https://www.archlinux.org/packages/extra/x86_64/nmap/)
0. `ykfde-ssh` from this PR

####  Basically what it does is
1. Server fires up network and dropbear
0. Client uses `ykfde-ssh Hostname` and it:
1.1 starts `ncat` to listen on `127.0.0.1:9000` and wait for the challenge
1.2 forwards `127.0.0.1:9000` to Server
0. Server uses `ss` to detect whether `127.0.0.1:9000` is open
0. Server uses `nc` to send the challenge to `127.0.0.1:9000` and obtains the response

I've tested it in a virtual machine and it seems to work pretty smoothly. However the configuration does tend to be overly complex and I actually wonder whether there is a better way of doing this rather than using SSH port forwarding. But @agherzan if you feel comfortable about this idea I can then make some time to tidy it up and submit a proper PR.

### CODE REVIEW

File: src/hooks/ykfde [hunk 1]
```diff
@@ -149,6 +150,8 @@ ykfde_do_it() {
   _rc=$?
 
   if [ "$_rc" -eq 0 ]; then
+    touch /.done
```
Comment (Vincent43) [inline src/hooks/ykfde]: I think those two should depend on `"$YKFDE_SSH" `

File: src/ykfde-ssh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+ncat -c 'read challenge; /bin/ykchalresp -2 "$challenge";' -l 9000 &
```
[the remark this answers is not in the document]
Comment (AmShaegar13) [reply src/ykfde-ssh]: I am pretty sure that this is single quoted intentionally.

[the remark this answers is not in the document]
Comment (agherzan) [reply src/ykfde-ssh]: Yes - it was to avoid expand. But fi that is the case, a comment and a shellcheck disable should be added.

**Comment by agherzan** [comment]
Shall we close this PR for now?

**Comment by Frederick888** [comment]
Ah, sorry, got busy for a while then completely forgot about this. I still would like to finish this feature, but I can't give you a time frame right now. Is it ok? Or I'm also happy to submit another PR when I actually manage to wrap it up.

**Comment by Vincent43** [comment]
Take your time, we can wait 😄

**Comment by agherzan** [comment]
That's absolutely alright. Just wanted to make sure it's still planned sometime in the future.

**Comment by Piraty** [comment]
https://github.com/dracut-crypt-ssh/dracut-crypt-ssh/ does that
